colexec: remove internal cancellation behavior from unordered synchronizer

asubiotto opened the pull request:
Previously, the unordered synchronizer would cancel all inputs if one of the
inputs encountered an error. This would result in possible context cancellation
errors racing with the original error and would sometimes cause the original
error to be overwritten (according to priority) in the distsql receiver (the
root of a query).

This behavior was incorrect, because what should happen is that the original
error should be propagated followed by a call to DrainMeta by the caller to
drain and close the remaining inputs. This commit removes internal context
cancellation in favor of this behavior.

Release note (bug fix): unexpected context cancellation errors could sometimes
be returned in the vectorized execution engine. This is now fixed.

Fixes #51647 
Fixes #52057

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52463)
<!-- Reviewable:end -->

asubiotto commented:
One concern I had was what if one of the inputs is blocked in `Next`, say, reading from a GRPC stream? However, I can't come up with a scenario in which this would be the case but it might be good to think more about this.

yuzefovich reviewed (commented):
LGTM, but is the race in the test concerning?

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-:-ME3Vp765ObzWs04uU0z:bhz7ktf)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

yuzefovich commented:
I think we should get this in sooner rather than later.

asubiotto commented:
You're right, thanks for the ping. The race was concerning but not something that could've happened before. Since we weren't waiting for inputs to exit in `Next`, `DrainMeta` had to deal with live inputs which resulted in this race. The new version properly synchronizes accesses using the existing channels and waitgroup.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-:-MEhLdtW7wuRQu8fAi_Y:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto)

___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 156 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-MEhKoBu-SNI3hbGCKHk:-MEhKoBu-SNI3hbGCKHl:bb36skz) ([raw file](https://github.com/cockroachdb/cockroach/blob/67c74ff695aeec1ee331d8fb43ae09362d2204f7/pkg/sql/colexec/parallel_unordered_synchronizer.go#L156)):*
> ```Go
> 		// batchCh is a buffered channel in order to offer non-blocking writes to
> 		// input goroutines. During normal operation, this channel will have at most
> 		// inputs messages. However, during DrainMeta, inputs might need to push
> ```

nit: `s/inputs messages/len(inputs) messages/g`.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 358 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-MEhLHGV5ERHqBtfQS43:-MEhLHGV5ERHqBtfQS44:b-gvkgbh) ([raw file](https://github.com/cockroachdb/cockroach/blob/67c74ff695aeec1ee331d8fb43ae09362d2204f7/pkg/sql/colexec/parallel_unordered_synchronizer.go#L358)):*
> ```Go
> 	}
> 
> 	// Unblock any goroutines currently waiting to be told to read a next batch,
> ```

nit: `s/a next/the next/g`.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Fixed the CI failures. One is an edge case described in `DrainMeta` and the other is related to the recent change to send metadata messages with errors on all streams. `TestVectorizedFlowShutdown` failed because it got multiple metadata messages for the same id. I'm not sure why this wasn't failing before, but I think it's probably got to do with the unordered synchronizer properly forwarding the rest of the metadata now in case of error, so I change the test expectations.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-:-MEhn-L-CQ6XkdxkwunR:b-jxaq8n)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52463)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto)

___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-MEiT2PF-C8Jasi7djMO:-MEiT2PF-C8Jasi7djMP:b-wrgqo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92ef26f31f339bbd326e3d76666befae2b14db0/pkg/sql/colexec/parallel_unordered_synchronizer.go#L363)):*
> ```Go
> 	// retrieve the next batch but encounters an error. There are now n+1 messages
> 	// in batchCh. Notifying all these inputs to read the next batch would result
> 	// in 2n+1 messages on batchCh, which would cause a deadlock since this
> ```

Given 2n+1 here should the creation of `batchCh` be updated?
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 371 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-MEiTB1J7A8SGQi45ZYx:-MEiTB1J7A8SGQi45ZYy:b-ovvzz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92ef26f31f339bbd326e3d76666befae2b14db0/pkg/sql/colexec/parallel_unordered_synchronizer.go#L371)):*
> ```Go
> 		select {
> 		case msg := <-s.batchCh:
> 			if msg == nil {
> ```

Do we ever send `nil` message?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
I don't believe the latest race failure was caused by this PR and submitted #52890 to track. Rerunning the build.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-:-MEvXwnJACbhJ4j8ZNpo:b-9mpia4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-MEiT2PF-C8Jasi7djMO:-MErk0AE5HnI-slP51Dr:b5z5751) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92ef26f31f339bbd326e3d76666befae2b14db0/pkg/sql/colexec/parallel_unordered_synchronizer.go#L363)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Given 2n+1 here should the creation of `batchCh` be updated?
</blockquote></details>

I would say no, because I think it's easier to reason about `2n` during normal operations and simply handle the complexity of an extra message here in `DrainMeta`.
___
*[pkg/sql/colexec/parallel_unordered_synchronizer.go, line 371 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52463#-MEiTB1J7A8SGQi45ZYx:-MErkAM_3O3E0tp2OXZn:b30q9kt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d92ef26f31f339bbd326e3d76666befae2b14db0/pkg/sql/colexec/parallel_unordered_synchronizer.go#L371)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Do we ever send `nil` message?
</blockquote></details>

We don't explicitly send `nil`, it's just the signal that the channel API gives us for when a channel is `close`d.


<!-- Sent from Reviewable.io -->

Commit:
colexec: remove internal cancellation behavior from unordered synchronizer

Previously, the unordered synchronizer would cancel all inputs if one of the
inputs encountered an error. This would result in possible context cancellation
errors racing with the original error and would sometimes cause the original
error to be overwritten (according to priority) in the distsql receiver (the
root of a query).

This behavior was incorrect, because what should happen is that the original
error should be propagated followed by a call to DrainMeta by the caller to
drain and close the remaining inputs. This commit removes internal context
cancellation in favor of this behavior.

Release note (bug fix): unexpected context cancellation errors could sometimes
be returned in the vectorized execution engine. This is now fixed.

asubiotto commented:
bors r=yuzefovich
